appserver/connectors/connectors-runtime/src/test/java/com/sun/enterprise/resource/pool/datastructure/RWLockDataStructureTest.java

```diff
@@ -76,25 +86,240 @@ public void raceConditions() throws Exception {
+            () -> assertThat("Free List Size", dataStructure.getFreeListSize(), equalTo(resourceCount + 100))
+        );
+
+        List<ResourceHandle> allResources = dataStructure.getAllResources();
+        assertThat("Resources Size", allResources, hasSize(dataStructure.getResourcesSize()));
+        for (ResourceHandle resource : allResources) {
+            assertThat(Collections.frequency(allResources, resource), equalTo(1));
+        }
+
+        threadPool.shutdownNow();
+    }
+
+    @Test
+    public void testAddResourceWithException() throws Exception {
+
+        handler = createNiceMock(ResourceHandler.class);
+        allocator = createNiceMock(ResourceAllocator.class);
+
+        ResourceHandle resource = createMockBuilder(ResourceHandle.class)
+                .withConstructor(Object.class, ResourceSpec.class, ResourceAllocator.class, ClientSecurityInfo.class)
+                .withArgs(null, null, null, null)
+                .createNiceMock();
+
+        expect(handler.createResource(allocator)).andThrow(new PoolingException());
+        expect(allocator.createResource()).andThrow(new PoolingException());
+        expect(handler.createResource(allocator)).andReturn(resource);
+        expect(allocator.createResource()).andReturn(resource);
+
+        replay(resource, handler, allocator);
+
+        DataStructure dataStructure = new RWLockDataStructure(null, 1, handler, null);
+
+        assertAll(
+            () -> assertThrows(PoolingException.class, () -> dataStructure.addResource(allocator, 1)),
+            () -> assertThat("Resources Size", dataStructure.getResourcesSize(), equalTo(0)),
+            () -> assertThat("Free List Size", dataStructure.getFreeListSize(), equalTo(0))
+         );
+
+        assertAll(
+            () -> assertThat("Add Resource", dataStructure.addResource(allocator, 1), equalTo(1)),
+            () -> assertThat("Resources Size", dataStructure.getResourcesSize(), equalTo(1)),
+            () -> assertThat("Free List Size", dataStructure.getFreeListSize(), equalTo(1))
+        );
+    }
+
+    @RepeatedTest(20)
+    @Timeout(value = 10, threadMode = ThreadMode.SEPARATE_THREAD)
+    public void testGetResource() throws Exception {
+
+        DataStructure dataStructure = new RWLockDataStructure(null, RESOURCE_COUNT, handler, null);
+
+        assertThat("Add Resources", dataStructure.addResource(allocator, RESOURCE_COUNT), equalTo(RESOURCE_COUNT));
+
+        List<Callable<ResourceHandle>> tasks = new ArrayList<>(TASK_COUNT);
+        for (int i = 0; i < TASK_COUNT; i++) {
+            tasks.add(dataStructure::getResource);
+        }
+
+        ExecutorService threadPool = Executors.newFixedThreadPool(THREAD_COUNT);
+
+        List<Future<ResourceHandle>> futures = threadPool.invokeAll(tasks);
+        assertAll(
+            () -> assertAll(futures.stream().map(f -> (Executable) f::get).toList()),
```

Comment:
@avpinchuk Hi, There seems to be a minor error here. it should be `.collect(Collectors.toList())` since GlassFish 7 still to be tested on JDK11+.

Reply:
Thank you very much for feedback! I'll fix it tonight.

Reply:
Fixed